docs(framework): ONNX guide (#2695)
bentoml/BentoML: docs/source/frameworks/onnx.rst

====
ONNX
====

Preface

> -------

Reviewer: Maybe we can give one-two sentence overviewing about what ONNX is and link it to the ONNX documentation page.

It is also worth to mention the Compatibility layer with ONNX. What is the lowest version that we support, and what runtime library we are supporting

Author: we have list minimum version requirements in notes?

Reviewer: Have you tested out with ONNX 1.7?

Author: I have tested 1.9. Any significancy  for 1.7?

Reviewer: 1.8 is the introduction of tensorrt, 1.7 has some bug related. Maybe we can just have a compatiblity section saying bentoml supported onnxruntime 1.8 and up.

In this tutorial, we will show how to train a model with PyTorch
TensorFlow and Scikit Learn, then we will convert resulting model to
ONNX format, use BentoML to save it and create a prediction service.

.. note::

   BentoML currently only support `ONNX Runtime
   <https://onnxruntime.ai>`_ as ONNX backend. BentoML requires either
   ``onnxruntime>=1.9`` or ``onnxruntime-gpu>=1.9`` to be installed.

.. seealso::

   ONNX support multiple machine learning frameworks and each
   framework may have unique requirements/recommendations for
   converting its own model format to ONNX format. Please refer to

>    resources listed below for in-depth knowledge of each framework.

Reviewer: Move this up before preface, and remove the `seealso` tag.

.. tab:: PyTorch

   - `Quick tutorial about exporting a model from PyTorch to ONNX <https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html>`_ from official PyTorch documentation.
   - `torch.onnx <https://pytorch.org/docs/stable/onnx.html>`_ from official PyTorch documentation. Pay special attention to section **Avoiding Pitfalls**, **Limitations** and **Frequently Asked Questions**.

.. tab:: TensorFlow

   TODO

.. tab:: Scikit Learn

   TODO

> 

Reviewer: There should be a way to do a sync tab.

Author: currently the tabs are synced (which means if you click tensorflow, all tabs turn to tensorflow, is this what you mean?)

      batch_size = 1 # can be any number
      # Tracing input to the model
      x = torch.randn(batch_size, 1, 224, 224, requires_grad=True)

      # Export the model
      torch.onnx.export(torch_model,
			x,
			"super_resolution.onnx",   # where to save the model (can be a file or file-like object)
			export_params=True,        # store the trained parameter weights inside the model file
			opset_version=10,          # the ONNX version to export the model to
			do_constant_folding=True,  # whether to execute constant folding for optimization
			input_names = ['input'],   # the model's input names
			output_names = ['output'], # the model's output names
			dynamic_axes={'input' : {0 : 'batch_size'},    # variable length axes
				      'output' : {0 : 'batch_size'}})

   Now we can compute the output using ONNX Runtime’s Python APIs:

   .. code-block:: python

      import onnxruntime

      ort_session = onnxruntime.InferenceSession("super_resolution.onnx")
      # compute ONNX Runtime output prediction
      ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(x)}
      # ONNX Runtime will return a list of outputs
      ort_outs = ort_session.run(None, ort_inputs)

>       print(ort_outs[0])

Reviewer: For this part, lets just provide a block of code how to create a pytorch model and convert it to onnx, and link it to the pytorch documentation for more information.

Author: That’s exactly what this part do? Do you thinks the SuperResolutionNet code is too long? or you want to remove ort_session part codes?

Reviewer: L99 to 105 is not needed

.. tab:: TensorFlow

   TODO

.. tab:: Scikit Learn

   TODO

> 

Reviewer: Lets provide a section called
```
Building a Service using ONNX
```

Reviewer: that section should also include a quick snippets on how user should include dependencies under a bentofile.yaml.

Author: will do

> Saving ONNX model with BentoML

Author: I'm not sure following several sections still need pytorch/tensorflow/sklearn tabs because our bentoml.onnx will just handle the `onnx` format and nothing is framework specified. If we still have three tabs then user may think bentoml.onnx will handle pytorch/tensorflow/sklearn differently.

Author: cc @parano

Reviewer: imo it is more about showing user how to do it right? having the major 3 framework support sounds like a good idea to me.

-----------------------------

To quickly save an ONNX model to BentoML's :ref:`Model
Store<concepts/model:Managing Models>`, first use ``onnx.load`` to

> load the exported ONNX model back into ``onnx.ModelProto`` object,

Reviewer: would be nice if you can reference the `onnx.ModelProto` to the API documentation.

Reviewer: I feel like save model should be a small section under the first block `Train a model and export to ONNX format` and no need to get into how batchable is disabled by default. The general tutorial should already covered that.

Author: I still think for framework support batching, we should talk about `batchable` because the default `batchable=False` may not result optimized performance @aarnphm

Reviewer: noted.

   }
   bentoml.onnx.save_model("super_resolution", onnx_model, signatures=signatures)

which will result:

.. code-block:: bash

   Model(tag="super_resolution:lwqr7ah5ocv3rea3", path="~/bentoml/models/super_resolution/lwqr7ah5ocv3rea3/")

.. note::

   ``save_model`` will use ``{"run": {"batchable": False}}`` as
   default signatures if ``signatures`` is not provided. Set
   ``batchable`` to ``False`` will disable BentoML's
   :ref:`guides/batching:Adaptive Batching` functionality. That's why
   we provide our own signatures here. Read more about :ref:`Model
   Signatures <concepts/model:Model Signatures>` and :ref:`Batch Input
   <concepts/model:Batching>`

.. seealso::

   ``save_model`` also has some :ref:`general options
   <concepts/model:Save A Trained Model>` for functionalities like
   saving metadata and custom objects.

Loading an ONNX model with BentoML
----------------------------------

We can use ``load_model`` to load an ONNX model back to memory:

.. code-block:: python

   ort_session = bentoml.onnx.load_model("onnx_super_resolution")

.. note::

   BentoML will load an ONNX model back as an
   ``onnxruntime.InferenceSession`` object which is ready to do

>    inference

Reviewer: Maybe it is nice to have one, two sentences separating the difference between `onnx` and `onnxruntime`.

Author: I think we can do that in the preface

Then we can do the inference:

.. code-block:: python

   test_input = np.random.randn(2, 1, 244, 244) # can accept arbitrary batch size
   ort_session.run(None, {"input": test_input.astype(np.float32)})

.. note::

   In above codes we need explicitly to convert input ndarray to
   float32 because ``onnxruntime.InferenceSession`` only expects
   single floats. In the following section we will see that BentoML
   runner will automatically cast input data to this type

In addition to :ref:`general options <concepts/model:Retrieve a saved
model>` provided by :code:`load_model`, you can pass ``providers`` to
set ONNX Runtime's `Execution Providers
<https://onnxruntime.ai/docs/execution-providers/>`_. You can also
pass ``session_options`` to set ``onnxruntime.InferenceSession``'s
`SessionOptions

> <https://onnxruntime.ai/docs/api/python/api_summary.html#sessionoptions>`_.

Reviewer: I think this can be its own sections about execution providers. And what is the default behaviour for runners with regarding this options.
You can get into what users can expect when a ONNX runner is running on which hardware.

.. code-block:: python

   ort_session = bentoml.onnx.load_model(
       "super_resolution",
       providers=["TensorrtExecutionProvider", "CUDAExecutionProvider", "CPUExecutionProvider"]

>    )

Reviewer: We want to focus user on Runner concept, hence the `load_model` can just be a small section. You can say something like:
```
To verify the saved model can be loaded properly:
...
```

We can still keep the note about converting the input to ndarray. its nice.

Author: I think we can remove the part after (including)
> In addition to :ref:`general options <concepts/model:Retrieve a saved model>` provided by :code:`load_model`...

Using Runners
-------------

.. seealso::

   :ref:`Runners<concepts/runner:Using Runners>` for more information on what is
   a Runner and how to use it.

> .. seealso::
> 
>    :ref:`Runners<concepts/runner:Specifying Required Resources>` on how to
>    provide options for a runner.

Reviewer: ```suggestion
```

To use ``onnx`` runner locally, access the model via ``get`` and
convert it to a runner:

.. code-block:: python

   test_input = np.random.randn(2, 1, 244, 244)

   runner = bentoml.onnx.get("super_resolution").to_runner()

   runner.init_local()

   runner.run.run(test_input)

.. note::

   You don't need to cast your input ndarray to ``np.float32`` for

>    runner input

Reviewer: I think we can remove this part in light of a Service example with ONNX.

Reviewer: Also a good note about the `runner.run.run` behaviour and why it named that way.

Like ``load_model``, you can customize ``providers`` and
``session_options`` when you create a runner:

.. code-block:: python

   providers=["TensorrtExecutionProvider", "CUDAExecutionProvider", "CPUExecutionProvider"]

   runner = bentoml.onnx.get("super_resolution").with_options(providers=providers).to_runner()

   runner.init_local()

Building a Service for **ONNX**
-------------------------------

.. seealso::

   :ref:`Building a Service <concepts/service:Service and APIs>` for how to
   create a prediction service with BentoML.

Miscellaneous Notes

> -------------------

Reviewer: No need to hide it under Misc note, each of the below section can live as its own content section.

.. tab:: PyTorch

   For PyTorch models, you can do that by specifying ``dynamic_axes``
   when using ``torch.onnx.export``

   .. code-block:: python

      torch.onnx.export(torch_model,
			x,
			"super_resolution.onnx",   # where to save the model (can be a file or file-like object)
			export_params=True,        # store the trained parameter weights inside the model file
			opset_version=10,          # the ONNX version to export the model to
			do_constant_folding=True,  # whether to execute constant folding for optimization
			input_names = ['input'],   # the model's input names
			output_names = ['output'], # the model's output names
			dynamic_axes={'input' : {0 : 'batch_size'},    # variable length axes
				      'output' : {0 : 'batch_size'}})

.. tab:: TensorFlow

   TODO

.. tab:: Scikit Learn

   TODO

Default Execution Providers Settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* When a CUDA compatible GPU is available, BentoML runner will use ``["CUDAExecutionProvider", "CPUExecutionProvider"]`` as the default Execution Providers.
* When CUDA compatible GPU is not available, BentoML runner will use
  ``["CPUExecutionProvider"]`` as the default Execution Providers.

> * You can override this setting using ``with_options`` when creating

Reviewer: lets provide an example here as well.

   create a prediction service with BentoML.

.. code-block:: python

   import bentoml

   import numpy as np
   from PIL import Image as PIL_Image
   from PIL import ImageOps
   from bentoml.io import Image

   runner = bentoml.onnx.get("super_resolution:latest").to_runner()

   svc = bentoml.Service("super_resolution", runners=[runner])

   @svc.api(input=Image(), output=Image())
   def sr(img) -> np.ndarray:
       img = img.resize((224, 224))
       gray_img = ImageOps.grayscale(img)
       arr = np.array(gray_img) / 255.0 # convert from 0-255 range to 0.0-1.0 range
       arr = np.expand_dims(arr, (0, 1)) # add batch_size, color_channel dims
       sr_arr = runner.run.run(arr)
       sr_arr = np.squeeze(sr_arr) # remove batch_size, color_channel dims
       sr_img = PIL_Image.fromarray(np.uint8(sr_arr * 255) , 'L')
       return sr_img

.. note::

   In above codes we use ``runner.run.run(input_data)`` to do
   inference. The first ``run`` is referring to
   ``onnxruntime.InferenceSession``'s ``run`` method, while the second
   ``run`` is BentoML's naming convention for doing runner inference
   for a model method. For example, for a Keras model with ``predict``

>    method, we will call ``runner.predict.run(input_data)``.

Author: @aarnphm Do you think this explains `runner.run.run` clearly enough?

Reviewer: ```suggestion
.. note::

   In above example, notice that there are two :code:`run` in ``runner.run.run(input_data)`` while running inference. The distinction between the two ``run`` are as follow:
   1.  The first ``run`` refers  to``onnxruntime.InferenceSession``'s ``run`` method, which is the API from onnxruntime to run `inference <https://onnxruntime.ai/docs/api/python/api_summary.html#data-inputs-and-outputs>`_.
   2. The second ``run`` refers to BentoML's runner inference API for invoking a model's signature. In the case of ONNX, it happens to have the same name as the ``InferenceSession`` endpoint.
```
@ssheng does this make sense?

Author: Aaron's modification looks more clear than mine so I will commit this suggestion (though I still think it's better to keep the keras's example because runner.predict.run() will give user a comparison). cc @ssheng @aarnphm

Reviewer: I think the keras distinction example should be left at a runners level docs. Let me draft a docs that talk dissect this components in a different docs pr.

Reviewer 2: I think having another example would be helpful. I'm slowly leaning towards including more information in the framework guides than less.

====
ONNX
====

> Preface
> -------

Reviewer: No need for this line
```suggestion
```

ONNX is an open format built to represent machine learning models. ONNX provides `high interoperability <https://onnx.ai/supported-tools.html#buildModel>`_  among various frameworks, as well as enable machine learning practitioners to maximize models' performance across `different hardware <https://onnx.ai/supported-tools.html#deployModel>`_.

Due to its high interoperability among frameworks, we recommend you to check out the framework integration with ONNX as it will contain specific recommendation and requirements for that given framework.

.. tab:: PyTorch

   - `Quick tutorial about exporting a model from PyTorch to ONNX <https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html>`_ from official PyTorch documentation.
   - `torch.onnx <https://pytorch.org/docs/stable/onnx.html>`_ from official PyTorch documentation. Pay special attention to section **Avoiding Pitfalls**, **Limitations** and **Frequently Asked Questions**.

.. tab:: TensorFlow

   - `tensorflow-onnx (tf2onnx) <https://github.com/onnx/tensorflow-onnx>`_ documentation.

> .. tab:: Scikit Learn

Reviewer: lets make all the tab in here https://sphinx-design.readthedocs.io/en/furo-theme/tabs.html#synchronised-tabs

   .. code-block:: python

      import torch.nn as nn
      import torch.nn.init as init

      class SuperResolutionNet(nn.Module):
	  def __init__(self, upscale_factor, inplace=False):
	      super(SuperResolutionNet, self).__init__()

	      self.relu = nn.ReLU(inplace=inplace)
	      self.conv1 = nn.Conv2d(1, 64, (5, 5), (1, 1), (2, 2))
	      self.conv2 = nn.Conv2d(64, 64, (3, 3), (1, 1), (1, 1))
	      self.conv3 = nn.Conv2d(64, 32, (3, 3), (1, 1), (1, 1))
	      self.conv4 = nn.Conv2d(32, upscale_factor ** 2, (3, 3), (1, 1), (1, 1))
	      self.pixel_shuffle = nn.PixelShuffle(upscale_factor)

	      self._initialize_weights()

	  def forward(self, x):
	      x = self.relu(self.conv1(x))
	      x = self.relu(self.conv2(x))
	      x = self.relu(self.conv3(x))
	      x = self.pixel_shuffle(self.conv4(x))
	      return x

	  def _initialize_weights(self):
	      init.orthogonal_(self.conv1.weight, init.calculate_gain('relu'))
	      init.orthogonal_(self.conv2.weight, init.calculate_gain('relu'))
	      init.orthogonal_(self.conv3.weight, init.calculate_gain('relu'))
	      init.orthogonal_(self.conv4.weight)

      torch_model = SuperResolutionNet()

>    For this tutorial, we will download some pre-trained weights. Note
>    that this model was not trained fully for good accuracy and is used
>    here for demonstration purposes only.
> 
>    .. code-block:: python

Reviewer: ```suggestion
```

No need to explain about this here. We maybe can just link to the super resolution tutorial. It should also use a pretrained model right?

Author: I think these sentences is for user who really followed our guides and then find out the result is suboptimal, then they will know it's not their faults (bad copy and paste) or our faults. @aarnphm

      # Load pretrained model weights
      model_url = 'https://s3.amazonaws.com/pytorch/test_data/export/superres_epoch100-44c6958e.pth'

      # Initialize model with the pretrained weights
      map_location = lambda storage, loc: storage
      if torch.cuda.is_available():
	  map_location = None
      torch_model.load_state_dict(model_zoo.load_url(model_url, map_location=map_location))

      # set the model to inference mode
      torch_model.eval()

   Exporting a model to onnx in PyTorch works via tracing or
   scripting. In this tutorial we will export a model using
   tracing. Note how we export the model with an input of
   ``batch_size=1``, but then specify the first dimension as dynamic
   in the ``dynamic_axes`` parameter in ``torch.onnx.export()``. The
   exported model will thus accept inputs of size ``[batch_size, 1,

>    224, 224]`` where ``batch_size`` can vary among each inference.

Reviewer: thoughts: Does this mean `bentoml.torchscript` is better used for this example?

Author: @aarnphm It's not about bentoml, the `torch.onnx.export` just has similar api like torchscript export

Reviewer: noted.

   .. code-block:: python

      import onnxruntime

      ort_session = onnxruntime.InferenceSession("super_resolution.onnx")
      # compute ONNX Runtime output prediction
      ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(x)}
      # ONNX Runtime will return a list of outputs
      ort_outs = ort_session.run(None, ort_inputs)
      print(ort_outs[0])

.. tab:: TensorFlow

   First let's install `tf2onnx <https://github.com/onnx/tensorflow-onnx>`_

   .. code-block:: bash

      pip install tf2onnx

   For this tutorial we will download a pretrained ResNet-50 model:

   .. code-block:: python

      import tensorflow as tf
      from tensorflow.keras.applications.resnet50 import ResNet50

      model = ResNet50(weights='imagenet')

   Then we can export the model to ONNX format. Notice that we use
   ``None`` in `TensorSpec
   <https://www.tensorflow.org/api_docs/python/tf/TensorSpec>`_ to
   denote the first input dimension as dynamic batch axies, which
   means this dimension can accept arbitrary input size.

   .. code-block:: python

      spec = (tf.TensorSpec((None, 224, 224, 3), tf.float32, name="input"),)

>       onnx_model, _ = tf2onnx.convert.from_keras(model, input_signature=spec, opset=13)

Reviewer: you can put this in one line. and run `black` to format the code.

Reviewer: But we should do this after 1.0. for docs cleanup and style compliant for docs.
cc @ssheng

Author: noted

	  return sr_img

.. tab:: TensorFlow

   .. code-block:: python

      import bentoml

      import numpy as np
      from bentoml.io import Image
      from bentoml.io import JSON

      runner = bentoml.onnx.get("onnx_resnet50:latest").to_runner()

      svc = bentoml.Service("onnx_resnet50", runners=[runner])

      @svc.api(input=Image(), output=JSON())
      def predict(img):

	  from tensorflow.keras.applications.resnet50 import preprocess_input, decode_predictions

	  img = img.resize((224, 224))
	  arr = np.array(img)
	  arr = np.expand_dims(arr, axis=0)
	  arr = preprocess_input(arr)
	  preds = runner.run.run(arr)
	  return decode_predictions(preds, top=1)[0]

.. tab:: Scikit Learn

   TODO

.. note::

   In above codes we use ``runner.run.run(input_data)`` to do
   inference. The first ``run`` is referring to
   ``onnxruntime.InferenceSession``'s ``run`` method, while the second
   ``run`` is BentoML's naming convention for doing runner inference
   for a model method. For example, for a Keras model with ``predict``
   method, we will call ``runner.predict.run(input_data)``.

When constructing a :ref:`bentofile.yaml <concepts/bento:Bento Build Options>`,

> there are two ways to include ONNX as a dependency, via ``python`` or

Reviewer 2: Should we call this `pip` instead?

Reviewer: It is included via python field right?

``conda``:

.. tab-set::

   .. tab-item:: python

      .. code-block:: yaml

         python:
         - onnx
	 - onnxruntime

   .. tab-item:: conda

      .. code-block:: yaml

         conda:
           channels:
           - conda-forge
           dependencies:
           - onnx
	   - onnxruntime

Using Runners
-------------

.. seealso::

   :ref:`Runners<concepts/runner:Using Runners>` for more information on what is
   a Runner and how to use it.

To use ``onnx`` runner locally, access the model via ``get`` and
convert it to a runner:

.. code-block:: python

   test_input = np.random.randn(2, 1, 244, 244)

   runner = bentoml.onnx.get("super_resolution").to_runner()

   runner.init_local()

   runner.run.run(test_input)

.. note::

   You don't need to cast your input ndarray to ``np.float32`` for
   runner input

Like ``load_model``, you can customize ``providers`` and
``session_options`` when you create a runner:

.. code-block:: python

   providers=["TensorrtExecutionProvider", "CUDAExecutionProvider", "CPUExecutionProvider"]

   runner = bentoml.onnx.get("onnx_super_resolution").with_options(providers=providers).to_runner()

>    runner.init_local()

Reviewer: imo we don't need this part as we should always recommend user to use runner in a service context. A small section sayinig using `load_model` to test out the model inference similar to fastai should be good enough.

Preface
-------

> ONNX support multiple machine learning frameworks and each framework
> may have unique requirements/recommendations for converting its own

Reviewer: ```suggestion
ONNX is an open format built to represent machine learning models. ONNX provides `high interoperability <https://onnx.ai/supported-tools.html#buildModel>`_  among various frameworks, as well as enable machine learning practitioners to maximize models' performance across `different hardware <https://onnx.ai/supported-tools.html#deployModel>`_. 

Due to its high interoperability among frameworks, we recommend you to check out the framework integration with ONNX as it will contain specific recommendation and requirements for that given framework.

```

> model format to ONNX format. Please refer to resources listed below
> for in-depth knowledge of each framework.

Reviewer: ```suggestion
```

.. tab:: TensorFlow

   - `tensorflow-onnx (tf2onnx) <https://github.com/onnx/tensorflow-onnx>`_ documentation.

.. tab:: Scikit Learn

   TODO

> In this tutorial, we will show how to train a model (or use a
> pretrained model) with PyTorch TensorFlow and Scikit Learn, then we
> will convert resulting model to ONNX format, use BentoML to save it
> and create a prediction service.

Reviewer: ```suggestion
Converting model frameworks to ONNX format
-----------------------------------------------
```

* When CUDA compatible GPU is not available, BentoML runner will use
  ``["CPUExecutionProvider"]`` as the default Execution Providers.

> If dependencies are installed, using other Execution Providers like
> ``TensorrtExecutionProvider`` may increase the performance. You can

Reviewer: ```suggestion
If ``onnxruntime-gpu`` is installed, using ``TensorrtExecutionProvider`` may improve inference runtime. You can
```

override the default setting using ``with_options`` when creating the
runner:

.. code-block:: python

   providers=["TensorrtExecutionProvider", "CUDAExecutionProvider", "CPUExecutionProvider"]

   runner = bentoml.onnx.get("onnx_super_resolution").with_options(providers=providers).to_runner()

> You can read more about Execution Providers at ONNX Runtime's
> `official documentation
> <https://onnxruntime.ai/docs/execution-providers/>`_

Reviewer: ```suggestion
Read more about Execution Providers in `ONNX Runtime's documentation
<https://onnxruntime.ai/docs/execution-providers/>`_
```
